api/api_helpers.py

```diff
@@ -561,3 +565,120 @@ def get_t_stat(length):
     dof = length-1
     t_crit = np.abs(scipy.stats.t.ppf((.05)/2,dof)) # for two sided!
     return t_crit/np.sqrt(length)
+
+
+def get_geo(ip):
+    try:
+        ip_obj = ipaddress.ip_address(ip)
+        if ip_obj.is_private:
+            return('52.53721666833642', '13.424863870661927')
+    except ValueError:
+        return (None, None)
+
+    query = "SELECT ip_address, data FROM ip_data WHERE created_at > NOW() - INTERVAL '24 hours' AND ip_address=%s;"
+    db_data = DB().fetch_all(query, (ip,))
+
+    if db_data is not None and len(db_data) != 0:
+        return (db_data[0][1].get('latitude'), db_data[0][1].get('longitude'))
+
+    latitude, longitude = get_geo_ipapi_co(ip)
+
+    if latitude is False:
+        latitude, longitude = get_geo_ip_api_com(ip)
+    if latitude is False:
+        latitude, longitude = get_geo_ip_ipinfo(ip)
+
+    #If all 3 fail there is something bigger wrong
+    return (latitude, longitude)
+
+
+def get_geo_ipapi_co(ip):
+
+    response = requests.get(f"https://ipapi.co/{ip}/json/", timeout=10)
+    print(f"Accessing https://ipapi.co/{ip}/json/")
+    if response.status_code == 200:
+        resp_data = response.json()
+
+        if 'error' in resp_data or 'latitude' not in resp_data or 'longitude' not in resp_data:
+            return (None, None)
+
+        resp_data['source'] = 'ipapi.co'
+
+        query = "INSERT INTO ip_data (ip_address, data) VALUES (%s, %s)"
```

Comment:
refactor to have this outside of the vendor specific functions?

Reply:
I thought about this but the lines we could refactor are actually only 2 the `insert` which are 2 lines. So the refactor would be more lines of code and not make it a lot clearer. Would make refactoring the insert better. I could go either way thinking about it. I see pro and cons in both